File: backend/python/token_crypto.py
"""
token_crypto.py

Python decryption helper for AES-256-GCM encrypted OAuth tokens.
Mirrors the PHP TokenCrypto implementation in backend/token_crypto.php.

Format: enc:<version>:<base64(iv + tag + ciphertext)>
Key: 32-byte key from ENCRYPTION_KEY env var (64 hex chars)
"""

import os
import base64
from typing import Optional

_PREFIX = "enc:"
_VERSION = 1


def _get_key() -> bytes:
    """Load the 32-byte encryption key from ENCRYPTION_KEY env var."""
    raw = os.environ.get("ENCRYPTION_KEY", "")
    if not raw or raw == "CHANGE_ME":
        raise RuntimeError("ENCRYPTION_KEY is not configured.")
    key = bytes.fromhex(raw)
    if len(key) != 32:
        raise RuntimeError("ENCRYPTION_KEY must be a 64-character hex string (32 bytes).")
    return key


def is_encrypted(value: str) -> bool:
    """Check if a value has the encrypted prefix."""
    return value.startswith(_PREFIX)


def decrypt(value: str) -> str:
    """Decrypt an AES-256-GCM encrypted string."""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    if not value.startswith(_PREFIX):
        raise ValueError("Value is not encrypted (missing prefix).")

    body = value[len(_PREFIX):]
    colon_idx = body.index(":")
    version = int(body[:colon_idx])
    if version != _VERSION:
        raise ValueError(f"Unsupported encryption version: {version}")

    packed = base64.b64decode(body[colon_idx + 1:])
    if len(packed) < 12 + 16:  # IV + tag minimum
        raise ValueError("Invalid encrypted payload.")

    iv = packed[:12]
    tag = packed[12:28]
    ciphertext = packed[28:]

    aesgcm = AESGCM(_get_key())
    # AESGCM.decrypt expects tag appended to ciphertext
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plaintext.decode("utf-8")


def decrypt_if_needed(value: Optional[str]) -> Optional[str]:
    """Decrypt only if encrypted, return None/empty as-is."""
    if not value:
        return value
    if is_encrypted(value):
        return decrypt(value)
    return value


def encrypt(plaintext: str) -> str:
    """Encrypt a plaintext string. Returns prefixed ciphertext."""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    iv = os.urandom(12)
    aesgcm = AESGCM(_get_key())
    ct_with_tag = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    # AESGCM.encrypt returns ciphertext + tag (16 bytes) appended
    packed = iv + ct_with_tag
    return f"{_PREFIX}{_VERSION}:{base64.b64encode(packed).decode()}"


def encrypt_if_needed(value: str) -> str:
    """Encrypt only if not already encrypted."""
    if not value or is_encrypted(value):
        return value
    return encrypt(value)

File: backend/token_crypto.php
<?php
declare(strict_types=1);

/**
 * backend/token_crypto.php
 *
 * Authenticated encryption for OAuth tokens at rest.
 * Uses AES-256-GCM (authenticated encryption with associated data).
 *
 * Design:
 *  - Encryption key loaded from ENCRYPTION_KEY constant (set via .env)
 *  - Each encrypted value includes: version byte + 12-byte IV + ciphertext + 16-byte auth tag
 *  - Prefix 'enc:' distinguishes encrypted values from legacy plaintext
 *  - Supports future key rotation via version byte
 *  - Decrypt never trusts ciphertext without authentication
 *  - Plaintext kept in memory only for the shortest possible time
 */

class TokenCrypto {
    private const CIPHER = 'aes-256-gcm';
    private const IV_LENGTH = 12;
    private const TAG_LENGTH = 16;
    private const PREFIX = 'enc:';
    private const KEY_VERSION = 1;

    private string $key;

    public function __construct() {
        $rawKey = defined('ENCRYPTION_KEY') ? ENCRYPTION_KEY : '';
        if ($rawKey === '' || $rawKey === 'CHANGE_ME') {
            throw new RuntimeException('ENCRYPTION_KEY is not configured. Set it in your .env file.');
        }
        $this->key = hex2bin($rawKey);
        if ($this->key === false || strlen($this->key) !== 32) {
            throw new RuntimeException('ENCRYPTION_KEY must be a 64-character hex string (32 bytes).');
        }
    }

    /**
     * Encrypt a plaintext string. Returns prefixed ciphertext.
     */
    public function encrypt(string $plaintext): string {
        $iv = random_bytes(self::IV_LENGTH);
        $tag = '';
        $ciphertext = openssl_encrypt(
            $plaintext,
            self::CIPHER,
            $this->key,
            OPENSSL_RAW_DATA,
            $iv,
            $tag,
            '',
            self::TAG_LENGTH
        );
        if ($ciphertext === false) {
            throw new RuntimeException('Token encryption failed.');
        }
        // Format: enc:version:base64(iv + tag + ciphertext)
        $packed = $iv . $tag . $ciphertext;
        return self::PREFIX . self::KEY_VERSION . ':' . base64_encode($packed);
    }

    /**
     * Decrypt an encrypted string back to plaintext.
     * Returns plaintext or throws RuntimeException on failure.
     */
    public function decrypt(string $encoded): string {
        if (!str_starts_with($encoded, self::PREFIX)) {
            throw new RuntimeException('Value is not encrypted (missing prefix).');
        }
        $body = substr($encoded, strlen(self::PREFIX));
        $colonPos = strpos($body, ':');
        if ($colonPos === false) {
            throw new RuntimeException('Invalid encrypted format.');
        }
        $version = (int)substr($body, 0, $colonPos);
        if ($version !== self::KEY_VERSION) {
            throw new RuntimeException("Unsupported encryption version: {$version}");
        }
        $packed = base64_decode(substr($body, $colonPos + 1), true);
        if ($packed === false || strlen($packed) < self::IV_LENGTH + self::TAG_LENGTH) {
            throw new RuntimeException('Invalid encrypted payload.');
        }
        $iv = substr($packed, 0, self::IV_LENGTH);
        $tag = substr($packed, self::IV_LENGTH, self::TAG_LENGTH);
        $ciphertext = substr($packed, self::IV_LENGTH + self::TAG_LENGTH);

        $plaintext = openssl_decrypt(
            $ciphertext,
            self::CIPHER,
            $this->key,
            OPENSSL_RAW_DATA,
            $iv,
            $tag
        );
        if ($plaintext === false) {
            throw new RuntimeException('Token decryption failed (authentication error).');
        }
        return $plaintext;
    }

    /**
     * Check if a value is encrypted (has our prefix).
     */
    public static function isEncrypted(string $value): bool {
        return str_starts_with($value, self::PREFIX);
    }

    /**
     * Encrypt a value only if it is not already encrypted.
     * Safe for migration: existing encrypted values pass through unchanged.
     */
    public function encryptIfNeeded(string $value): string {
        if ($value === '' || self::isEncrypted($value)) {
            return $value;
        }
        return $this->encrypt($value);
    }

    /**
     * Decrypt a value if it is encrypted, or return as-is if plaintext.
     * WARNING: Prefer decrypt() for new code. This exists only for migration compatibility.
     */
    public function decryptIfNeeded(string $value): string {
        if ($value === '' || !self::isEncrypted($value)) {
            return $value;
        }
        return $this->decrypt($value);
    }
}

/**
 * Get or create a singleton TokenCrypto instance.
 * Throws if ENCRYPTION_KEY is not configured.
 */
function getTokenCrypto(): TokenCrypto {
    static $instance = null;
    if ($instance === null) {
        $instance = new TokenCrypto();
    }
    return $instance;
}
